R/infillOptFocus.R

```diff
@@ -8,11 +8,35 @@
 # See infillOptCMAES.R for interface explanation.
 infillOptFocus = function(infill.crit, models, control, par.set, opt.path, design, iter, ...) {
   global.y = Inf
+  
+  allRequirements = extractSubList(par.set$pars, "requires", simplify = FALSE)
+  allUsedVars = unique(do.call(base::c, sapply(allRequirements, all.vars)))
+  forbiddenVars = getParamIds(filterParams(par.set, type = c("discretevector", "logicalvector")))
+  if (any(allUsedVars%in% forbiddenVars)) {
+    stop("Cannot do focus search when some variables have requirements that depend on discrete or logical vector parameters.")
+  }
+  
 
-  # restart the whole crap some times
+  # perform multiple starts
   for (restart.iter in seq_len(control$infill.opt.restarts)) {
     # copy parset so we can shrink it
     ps.local = par.set
+    
+    # handle discrete vectors:
+    # The problem is that for discrete vectors, we can't adjust the values dimension-wise.
+    # Therefore, for discrete vectors, we always drop the last level and instead have a
+    # mapping that maps, for each discrete vector param and for each dimension, from 
+    # the sampled value (levels 1 to n - #(dropped levels)) to levels with random dropouts.
+    discreteVectorMapping = lapply(filterParams(par.set, type = c("discretevector", "logicalvector"))$pars,
```

Comment:
IMHO you can store `filterParams(par.set, type = c("discretevector", "logicalvector"))` as `ps.disic.log.vec` or something, as you use it twice here.